python/tank_vendor/shotgun_deploy/descriptor_io/factory.py

```diff
@@ -0,0 +1,170 @@
+    :param fallback_roots: List of immutable fallback cache locations where
+                           apps will be searched for. Note that when descriptors
+                           download new content, it always ends up in the
+                           bundle_cache_root.
+    :returns: Descriptor object
+    """
+    from .appstore import IODescriptorAppStore
+    from .dev import IODescriptorDev
+    from .path import IODescriptorPath
+    from .shotgun_entity import IODescriptorShotgunEntity
+    from .git import IODescriptorGit
+    from .git_branch import IODescriptorGitBranch
+    from .manual import IODescriptorManual
+
+    if isinstance(location, basestring):
+        # translate uri to dict
+        location_dict = location_uri_to_dict(location)
+    else:
+        location_dict = location
+
+    if location_dict.get("type") == "app_store":
+        descriptor = IODescriptorAppStore(location_dict, sg, descriptor_type)
+
+    elif location_dict.get("type") == "shotgun":
+        descriptor = IODescriptorShotgunEntity(location_dict, sg)
+
+    elif location_dict.get("type") == "manual":
+        descriptor = IODescriptorManual(location_dict)
+
+    elif location_dict.get("type") == "git":
+        descriptor = IODescriptorGit(location_dict)
+
+    elif location_dict.get("type") == "git_branch":
+        descriptor = IODescriptorGitBranch(location_dict)
+
+    elif location_dict.get("type") == "dev":
+        descriptor = IODescriptorDev(location_dict)
+
+    elif location_dict.get("type") == "path":
+        descriptor = IODescriptorPath(location_dict)
+
+    else:
+        raise ShotgunDeployError("Unknown location type for '%s'" % location_dict)
+
+    # specify where to go look for caches
+    descriptor.set_cache_roots(bundle_cache_root, fallback_roots)
+
+    if location_dict.get("version") == constants.LATEST_DESCRIPTOR_KEYWORD:
+        log.debug("Latest keyword detected. Searching for latest version...")
+        descriptor = descriptor.get_latest_version()
+        log.debug("Resolved latest to be %r" % descriptor)
+
+    return descriptor
+
+
+def location_uri_to_dict(location_uri):
+    """
+    Translates a location uri into a location dictionary, suitable for
+    use with the create_io_descriptor factory method below.
+
+    :param location_uri: location string uri
+    :returns: location dictionary
+    """
+    from .appstore import IODescriptorAppStore
+    from .dev import IODescriptorDev
+    from .path import IODescriptorPath
+    from .shotgun_entity import IODescriptorShotgunEntity
+    from .git import IODescriptorGit
+    from .git_branch import IODescriptorGitBranch
+    from .manual import IODescriptorManual
+
+    chunks = location_uri.split(constants.LOCATOR_URI_SEPARATOR)
+    if chunks[0] != constants.LOCATOR_URI_PREFIX or len(chunks) < 3:
+        raise ShotgunDeployError("Invalid uri %s" % location_uri)
+    descriptor_type = urllib.unquote(chunks[1])
+
+    if descriptor_type == "app_store":
```

Comment:
Registration system would be nice here as well. Each descriptor type I would image is registered with a `type` to match against and you could just ask the registry for the class with the matching type.